src/globus_sdk/experimental/globus_app/globus_app.py

```diff
@@ -0,0 +1,429 @@
+            driving login flows with a ``UserApp``. If not passed a
+            ``CommandLineLoginFLowManager`` will be used. A ``ClientApp`` will
+            error if this value is not None.
+        :param token_storage: a ``TokenStorage`` instance or class that will
+            be used for storing token data. If not passed a ``JSONTokenStorage``
+            will be used.
+        :param refresh_tokens: If True, the ``GlobusApp`` will request refresh tokens
+            for long lived access.
+        """
+        self.refresh_tokens = refresh_tokens
+        self.login_flow_manager = login_flow_manager
+        self.token_storage = token_storage
+
+
+class GlobusApp(metaclass=abc.ABCMeta):
+    """
+    ``GlobusApp` is an abstract base class providing an interface for simplifying
+    authentication with Globus services.
+
+    A ``GlobusApp`` manages scope requirements across multiple resource servers,
+    runs login flows through its ``LoginFlowManager``, handles token storage and
+    validation through its ``ValidatingTokenStorage``, and provides up-to-date
+    authorizers from its ``AuthorizerFactory`` through ``get_authorizer``.
+
+    In the near future, a ``GlobusApp`` will be accepted as an initialization parameter
+    to any Globus service client, automatically handling the client's default
+    scope requirements and providing the client an authorizer.
+    """
+
+    _login_client: AuthLoginClient
+    _authorizer_factory: AuthorizerFactory[GlobusAuthorizer]
+
+    def __init__(
+        self,
+        app_name: str,
+        *,
+        scope_requirements: dict[str, list[Scope]] | None = None,
+        config: GlobusAppConfig | None = None,
+    ):
+        """
+        :param app_name: A string to identify this app. Used for default tokenstorage
+            location and in the future will be used to set user-agent when this app is
+            attached to a service client
+        :param scope_requirements: A dict of lists of required scopes indexed by
+            their resource server.
+        :config: A ``GlobusAppConfig`` used to control various behaviors of this app.
+        """
+        self.app_name = app_name
+
+        if config:
+            self.config = config
+        else:
+            # default config
+            self.config = GlobusAppConfig()
+
+        self._initialize_login_client()
+
+        # get our initial scope requirements make sure we at least have "openid"
+        # to get identity information for token validation
+        auth_rs = str(AuthClient.resource_server)
+        openid_scope = Scope(AuthScopes.openid)
+        if scope_requirements:
+            if auth_rs not in scope_requirements:
+                scope_requirements[auth_rs] = [openid_scope]
+            else:
+                scope_requirements[auth_rs] = Scope.merge_scopes(
+                    scope_requirements[auth_rs], [openid_scope]
+                )
+            self._scope_requirements = scope_requirements
```

Comment:
Could we make scope_requirements public?

Keeping it private removes the ability for users to read what scope requirements have actually been defined on the App.

Reply:
In a similar vein, I'd recommend that we do scope merging on scope read (when initiating a login flow) not write (when a scope requirement is being added).

Scope deduplication (not adding the same exact scope twice) could make sense; but merging has the potential to mutate the change data structure from under the user unexpected ways.